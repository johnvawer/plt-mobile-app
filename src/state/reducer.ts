type ActionMap<M extends { [index: string]: any }> = {
  [Key in keyof M]: M[Key] extends undefined
    ? {
      type: Key
    }
    : {
      type: Key
      payload: M[Key]
    }
}

export enum Types {
  Add = 'ADD_PRODUCT',
  Remove = 'REMOVE_PRODUCT'
}

interface Product {
  id: number
  name: string
  price: number
  quantity: number
  totalPrice: number
}

interface ProductPayload {
  [Types.Add]: {
    id: number
    name: string
    price: number
  }
  [Types.Remove]: {
    id: number
  }
}

export type ProductActions = ActionMap<ProductPayload>[keyof ActionMap<ProductPayload>]

export const productReducer = (state: Product[], action: ProductActions): Product[] => {
  const itemExists = state.find(product => product.id === action.payload.id)

  switch (action.type) {
    case 'ADD_PRODUCT':
      if (itemExists != null) {
        return state.map(product =>
          product.id === itemExists.id
            ? {
              ...product,
              quantity: product.quantity + 1,
              totalPrice: product.price * (product.quantity + 1)
            } : product
        )
      }

      return [
        ...state,
        {
          id: action.payload.id,
          name: action.payload.name,
          price: action.payload.price,
          quantity: 1,
          totalPrice: action.payload.price
        }
      ]
    case 'REMOVE_PRODUCT':
      if ((itemExists != null) && itemExists.quantity > 1) {
        return state.map(product =>
          product.id === itemExists.id
            ? {
              ...product,
              quantity: product.quantity - 1,
              totalPrice: product.price * (product.quantity - 1)
            } : product
        )
      }

      return [
        ...state.filter(product => product.id !== action.payload.id)
      ]
    default:
      return state
  }
}
