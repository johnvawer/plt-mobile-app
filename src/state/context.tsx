import { createContext, Dispatch } from 'react'
import { productReducer, ProductActions } from './reducer'
interface ComputedProduct {
  id: number
  name: string
  price: number
  quantity: number
  totalPrice: number
}

interface InitialState {
  products: ComputedProduct[]
}

export const initialState = {
  products: []
}

export const AppContext = createContext<{
  state: InitialState
  dispatch: Dispatch<ProductActions>
}>({
  state: initialState,
  dispatch: () => null
})

export const mainReducer = ({ products }: InitialState, action: ProductActions): { products: ComputedProduct[] } => ({
  products: productReducer(products, action)
})
