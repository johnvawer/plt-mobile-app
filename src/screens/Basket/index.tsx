import React, { useContext, useState, useEffect } from 'react'
import { View, Text, FlatList, StyleSheet } from 'react-native'

import { AppContext } from '../../state/context'
import { Types } from '../../state/reducer'
import ProductQuantity from '../../components/Molecules/ProductQuantity'
import ProductPrice from '../../components/Atoms/ProductPrice'

interface BasketTotals {
  totalAmount: number
  totalItems: number
}

const Basket = (): JSX.Element => {
  const { state, dispatch } = useContext(AppContext)
  const { products } = state
  const [basketTotals, setBasketTotals] = useState<BasketTotals>({
    totalAmount: 0,
    totalItems: 0
  })

  useEffect(() => {
    if (products != null && (products.length > 0)) {
      const currentBasket = products.reduce((totals: BasketTotals, product: ComputedProduct) => {
        totals.totalAmount = totals.totalAmount += product.totalPrice
        totals.totalItems = totals.totalItems += product.quantity
        return totals
      }, {
        totalAmount: 0,
        totalItems: 0
      })

      setBasketTotals(currentBasket)
    }
  }, [products])

  if (products.length === 0) {
    return (
      <View>
        <Text>Your basket is empty</Text>
      </View>
    )
  }

  const onPressMore = (product: ComputedProduct): void => {
    const { id, name, price } = product

    dispatch({
      type: Types.Add,
      payload: {
        id,
        name,
        price
      }
    })
  }

  const onPressLess = (productId: number): void => {
    dispatch({
      type: Types.Remove,
      payload: {
        id: productId
      }
    })
  }

  return (
    <>
      <View style={{ flex: 0.75 }}>
        <FlatList
          data={products}
          renderItem={({ item }: { item: ComputedProduct }) => {
            return (
              <View style={styles.productRow}>
                <View>
                  <Text>Name: {item.name}</Text>
                  <Text>Price: <ProductPrice amount={item.price} /></Text>
                  <Text>Total Price: <ProductPrice amount={item.totalPrice} /></Text>
                </View>
                <View>
                  <ProductQuantity
                    onPressLess={() => onPressLess(item.id)}
                    onPressMore={() => onPressMore(item)}
                    quantity={item.quantity}
                  />
                </View>
              </View>
            )
          }}
        />
      </View>
      <View style={{ flex: 0.25 }}>
        <Text testID='totalItems'>Total Items: {basketTotals.totalItems}</Text>
        <Text>Total Amount: <ProductPrice testID='totalAmount' amount={basketTotals.totalAmount} /></Text>
      </View>
    </>
  )
}

const styles = StyleSheet.create({
  productRow: {
    margin: 10
  }
})

export default Basket
