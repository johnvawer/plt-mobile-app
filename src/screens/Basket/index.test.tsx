import React from 'react'
import { render, fireEvent } from '@testing-library/react-native'
import { AppContext } from '../../state/context'
import Basket from '.'

const basketProducts = [{
  id: 1,
  name: 'product one',
  price: 10,
  quantity: 1,
  totalPrice: 10
}, {
  id: 2,
  name: 'product two',
  price: 4,
  quantity: 6,
  totalPrice: 24
}, {
  id: 3,
  name: 'product three',
  price: 2.99,
  quantity: 2,
  totalPrice: 5.98
}]

describe('<Basket />', () => {
  let dispatchMock: jest.Mock
  let initialState: {
    state: { products: ComputedProduct[] }
    dispatch: () => void
  }

  beforeEach(() => {
    dispatchMock = jest.fn()
    initialState = {
      state: { products: [] },
      dispatch: dispatchMock
    }
  })

  afterEach(() => {
    dispatchMock.mockRestore()
  })

  it('should display an empty message if there are no products in the basket', () => {
    const { toJSON } = render(
      <AppContext.Provider value={initialState}>
        <Basket />
      </AppContext.Provider>
    )

    expect(toJSON()).toMatchSnapshot()
  })

  it('should display the correct basket totals', () => {
    initialState.state.products = [...basketProducts]

    const { getByTestId } = render(
      <AppContext.Provider value={initialState}>
        <Basket />
      </AppContext.Provider>
    )
    const totalAmount = getByTestId('totalAmount')
    const totalItems = getByTestId('totalItems')

    expect(totalItems.props.children[1]).toEqual(9)
    expect(totalAmount.props.children.join('')).toEqual('Â£39.98')
  })

  it('should dispatch to add a product to the store on pressing "+"', () => {
    initialState.state.products = [...basketProducts]

    const { getAllByText } = render(
      <AppContext.Provider value={initialState}>
        <Basket />
      </AppContext.Provider>
    )

    const plusButton = getAllByText('+')
    fireEvent.press(plusButton[0])

    expect(dispatchMock).toHaveBeenCalledTimes(1)
    expect(dispatchMock).toHaveBeenCalledWith({
      type: 'ADD_PRODUCT',
      payload: {
        id: 1,
        name: 'product one',
        price: 10
      }
    })
  })

  it('should dispatch to remove a product to the store on pressing "-"', () => {
    initialState.state.products = [...basketProducts]

    const { getAllByText } = render(
      <AppContext.Provider value={initialState}>
        <Basket />
      </AppContext.Provider>
    )

    const minusButton = getAllByText('-')
    fireEvent.press(minusButton[1])

    expect(dispatchMock).toHaveBeenCalledTimes(1)
    expect(dispatchMock).toHaveBeenCalledWith({
      type: 'REMOVE_PRODUCT',
      payload: {
        id: 2
      }
    })
  })
})
