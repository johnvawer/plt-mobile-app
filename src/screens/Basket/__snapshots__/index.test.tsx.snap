// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Basket /> should display an empty message if there are no products in the basket 1`] = `
<View>
  <Text>
    Your basket is empty
  </Text>
</View>
`;
